### PR TITLE
packages: fix up parse_accept_language()

### LINKED ISSUE #19526: Locale selection code in Python bridge is broken
### Explain what happens

Two issues:

1. We no longer properly return an empty document in the case that `en` is explicitly requested.  We rather notice that `po.en.js` doesn't exist, then move on to the next item in the list.  The effective result of that is that for a user who has `en, fr` as their languages list, we're going to end up picking French for them.  This is not the Canonical Desktop Team, so I guess that's not OK.
2. We don't treat fallbacks to "stripped" versions of locales in an RFC-compatible way.  The RFC says that if `de-de` is in the list we need to consider `de` before proceeding to the following items on the list.  We consider the stripped versions after searching the original list.  The effective result of that is that for a user who has `de-de, nl` as their languages list, we're going to end up giving them Dutch before German.
3. Maybe biggest issue of all: none of these problems were caught by tests.  Point 1 was a regression introduced in f4be9066c9d3f5dbad0316231b5ad98c117394cf.  We need to up our game here with some better unit and/or integration tests.

### Version of Cockpit

_No response_

### Where is the problem in Cockpit?

None

### Server operating system

None

### Server operating system version

_No response_

### What browsers are you using?

_No response_

### System log

_No response_

### PR DESCRIPTION
**Pull request by allisonkarlitskaya**
Make several fixes to our handling of the `Accept-Language:` header:

 - we used to crash if `q=` contained invalid values due to an unchecked cast with `float()`.  Catch that error and ignore the entry in that case.

 - our handling of English was incorrect.  We would look for and fail to find po.en.js, and move on to the next item in the list.  That means if the user listed English first, followed by another language which we did support, they'd see Cockpit in that other language, which is unexpected.  This is a regression introduced by f4be9066c9.  Now we drop all items that sort after English.

 - our handling of fallbacks (ie: 'de' from 'de-de') was incorrect. RFC4647 §3.4 says that a "Lookup" should consider items in the order they're found, stripping each item down to its base form, before considering the next item.  This passes a gut check, as well: a user who lists `de-de, nl` probably expects to see a German translation before a Dutch one.

We also now mark the parser code as `@lru_cache`.  This makes sense: within a given run of cockpit-bridge, we're likely to see very many copies of the Accept-Language header, and they're practically always going to be identical.  Make sure we now accept and return immutable types to prevent weird side-effects.

While we're at it, up our testing game a bit to make sure we don't mistreat requests for 'English' in the future.

Closes #19526

### CODE REVIEW

**Comment by allisonkarlitskaya** [comment]
Note: this change means that a language list like `en-ca, en-gb, en` is entirely pointless.  After `en-ca` is considered, `en` will be directly considered after it, effectively ending the search.  

It also means that lists like `de-at, de-ch` are almost pointless: after we fail to find `de-at` we'll immediately try to find `de` (which will typically be there).  `de-ch` won't be considered in that case.  This is the correct behaviour per RFC 4647.

**Comment by allisonkarlitskaya** [comment]
```
Traceback (most recent call last):
  File "/work/bots/make-checkout-workdir/test/verify/check-storage-stratis", line 869, in testBasic
    m.execute(f"stratis pool stop {self.stop_type_opt} pool0")
                                   ^^^^^^^^^^^^^^^^^^
AttributeError: 'TestStorageStratisNBDE' object has no attribute 'stop_type_opt'
```

hihi #19525

File: test/pytest/test_packages.py [hunk 1]
```diff
@@ -160,6 +162,31 @@ def test_condition_hides_priority(pkgdir):
...
+    assert '/javascript' in document.content_type
+    assert document.data.read() == b''
+
+    document = packages.load_path('/one/po.js', {'Accept-Language': 'de;q=0.9, fr;q=0.7, en'})
+    assert '/javascript' in document.content_type
+    assert document.data.read() == b''
+
+    document = packages.load_path('/one/po.js', {'Accept-Language': 'de;q=0.9, fr, en-ca'})
+    assert '/javascript' in document.content_type
+    assert document.data.read() == b''
```
Comment (martinpitt) [inline test/pytest/test_packages.py:187]: One interesting test case would be `en;q=0.9, de`. Intuitively that should result in "de", but given how my intuition was already fooled before by the "insert fallbacks with no country" before (which feels actively wrong for some cases), we better pin it down?

File: src/cockpit/packages.py [hunk 2]
```diff
@@ -62,28 +62,55 @@
...
+    logger.debug('parse_accept_language(%r)', accept_language)
+    locales_with_q = []
+    for entry in accept_language.split(','):
+        entry = entry.strip().lower()
+        logger.debug('  entry %r', entry)
+        locale, _, qstr = entry.partition(';q=')
+        try:
+            q = float(qstr or 1.0)
+        except ValueError:
+            continue  # ignore malformed entry
```
Comment (martinpitt) [inline src/cockpit/packages.py:88]: FTR, this handles the case of an empty AcceptLanguage header.

File: src/cockpit/packages.py [hunk 3]
```diff
@@ -62,28 +62,55 @@
...
-
-    return [locale for locale, _ in sorted(locales, key=lambda k: k[1], reverse=True)]
+    logger.debug('parse_accept_language(%r)', accept_language)
+    locales_with_q = []
+    for entry in accept_language.split(','):
+        entry = entry.strip().lower()
+        logger.debug('  entry %r', entry)
+        locale, _, qstr = entry.partition(';q=')
+        try:
+            q = float(qstr or 1.0)
```
Comment (martinpitt) [inline src/cockpit/packages.py:86]: I'm somewhat surprised that mypy accepts this `str | float` construction. But so much the better :grin:

File: src/cockpit/packages.py [hunk 4]
```diff
@@ -62,28 +62,55 @@
...
+        except ValueError:
+            continue  # ignore malformed entry
+
+        while locale:
+            logger.debug('    adding %r q=%r', locale, q)
+            locales_with_q.append((locale, q))
+            # strip off '-detail' suffixes until there's nothing left
+            locale, _, _region = locale.rpartition('-')
```
Comment (martinpitt) [inline src/cockpit/packages.py:94]: This feels weird. I've never seen a ll-cc-xx locale, what should that even be? Then again, it should handle weird cases like `--de` or `fr-fr-fr-----` in an acceptable way, so good enough.

File: test/pytest/test_packages.py [hunk 5]
```diff
@@ -160,6 +162,31 @@ def test_condition_hides_priority(pkgdir):
...
+
```
Comment (martinpitt) [inline test/pytest/test_packages.py:188]: It would be good to cover `""` (should result in "", i.e. implied English), and the invalid locales like `en-us-xx`, as you explicitly handle them in the loop above.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Very nice, thanks for fixing this! I'd like to see some more edge cases covered in tests, but code looks good!

File: src/cockpit/packages.py (hunk 2 above)
Comment (allisonkarlitskaya) [reply src/cockpit/packages.py:88]: Which part of it?  The `try:` not, because if the accept-language header is empty (or indeed, if no quality is specified) then `qstr` will be empty and we'll get a default value of `1.0`.  But note below: we only loop so long as `locale` is non-empty, which means we won't add the locale in that case.

File: src/cockpit/packages.py (hunk 4 above)
Comment (allisonkarlitskaya) [reply src/cockpit/packages.py:94]: [RFC4647](https://datatracker.ietf.org/doc/html/rfc4647#section-3.4) envisions this, giving elaborate examples like `zh-Hant-CN-x-private1-private2` and shows how it gets iteratively stripped down to `zh`.  I don't think these cases are very interesting to us, but if they do pop up, we'd be handling them properly (from my read).

File: test/pytest/test_packages.py (hunk 5 above)
Comment (allisonkarlitskaya) [reply test/pytest/test_packages.py:188]: OK.

File: src/cockpit/packages.py (hunk 2 above)
Comment (martinpitt) [reply src/cockpit/packages.py:88]: ah right, `''` is falsy, so it would fall back to 1.0. So yes, it's the while loop. I mostly just want to ensure that this is handled somehow, as it's not immediately obvious, and `''.split(',')` is (annoyingly) `['']`, not `[]`.

**Review by martinpitt — APPROVED** [review]
Cheers!
